From 2b102a1046630b072d80e36c872e2aa4e28052ca Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Fri, 7 Jan 2022 17:12:43 +0100
Subject: [PATCH] jack: remember last return from jack_get_buffer_size

Remember what we last returned from jack_get_buffer_size and only
emit a buffersize change event when somwthing new is configured.

Fixes startup of jconvolver.

Fixes #1989
---
 pipewire-jack/src/pipewire-jack.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/pipewire-jack/src/pipewire-jack.c b/pipewire-jack/src/pipewire-jack.c
index 595d786f6..f338d2469 100644
--- a/pipewire-jack/src/pipewire-jack.c
+++ b/pipewire-jack/src/pipewire-jack.c
@@ -1227,7 +1227,8 @@ do_buffer_frames(struct spa_loop *loop,
 {
 	uint32_t buffer_frames = *((uint32_t*)data);
 	struct client *c = user_data;
-	do_callback_expr(c, c->buffer_frames = buffer_frames, bufsize_callback, buffer_frames, c->bufsize_arg);
+	if (c->buffer_frames != buffer_frames)
+		do_callback_expr(c, c->buffer_frames = buffer_frames, bufsize_callback, buffer_frames, c->bufsize_arg);
 	recompute_latencies(c);
 	return 0;
 }
@@ -3985,6 +3986,7 @@ jack_nframes_t jack_get_buffer_size (jack_client_t *client)
 				res = c->position->clock.duration;
 		}
 	}
+	c->buffer_frames = res;
 	pw_log_debug("buffer_frames: %u", res);
 	return res;
 }
-- 
GitLab

